base/pkg/libgit2/types.jl

```diff
@@ -0,0 +1,193 @@
+    completion::Ptr{Void}
+    credentials::Ptr{Void}
+    certificate_check::Ptr{Void}
+    transfer_progress::Ptr{Void}
+    update_tips::Ptr{Void}
+    pack_progress::Ptr{Void}
+    push_transfer_progress::Ptr{Void}
+    push_update_reference::Ptr{Void}
+    payload::Ptr{Void}
+end
+RemoteCallbacksStruct() = RemoteCallbacksStruct(one(Cuint),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0),
+                                                Ptr{Void}(0))
+
+type CloneOptionsStruct
+    version::Cuint
+    checkout_opts::CheckoutOptionsStruct
+    remote_callbacks::RemoteCallbacksStruct
+    bare::Cint
+    localclone::Cint
+    checkout_branch::Ptr{UInt8}
+    signature::Ptr{Void}
+    repository_cb::Ptr{Void}
+    repository_cb_payload::Ptr{Void}
+    remote_cb::Ptr{Void}
+    remote_cb_payload::Ptr{Void}
+end
+
+CloneOptionsStruct() = CloneOptionsStruct(one(Cuint),
+                                          CheckoutOptionsStruct(),
+                                          RemoteCallbacksStruct(),
+                                          zero(Cint),
+                                          zero(Cint),
+                                          Ptr{UInt8}(0),
+                                          Ptr{Void}(0),
+                                          Ptr{Void}(0), Ptr{Void}(0),
+                                          Ptr{Void}(0), Ptr{Void}(0)
+                                        )
+
+# Common types
+for (typ, ref, fnc) in ((:GitRemote,     :Void, :(:git_remote_free)),
+                        (:GitRevWalker,  :Void, :(:git_revwalk_free)),
+                        (:GitConfig,     :Void, :(:git_config_free)),
+                        (:GitReference,  :Void, :(:git_reference_free)),
+                        (:GitDiff,       :Void, :(:git_diff_free)),
+                        (:GitIndex,      :Void, :(:git_index_free)),
+                        (:GitSignature,  :SignatureStruct, :(:git_signature_free)),
+                        (:GitRepo,       :Void, :(:git_repository_free)))
+    @eval type $typ
+        ptr::Ptr{$ref}
+        function $typ(ptr::Ptr{$ref})
+            @assert ptr != C_NULL
+            obj = new(ptr)
+            return obj
+        end
+    end
+
+    @eval function finalize(obj::$typ)
+        if obj.ptr != C_NULL
+            ccall(($fnc, :libgit2), Void, (Ptr{$ref},), obj.ptr)
+            obj.ptr = C_NULL
+        end
+    end
+
+    @eval Base.isempty(obj::$typ) = (obj.ptr == C_NULL)
+end
```

Comment:
Maybe some `AbstractGitObject` supertype for all wrappers.
```julia
abstract AbstractGitObject
type GitRepo <: AbstractGitObject
abstract GitObject <: AbstractGitObject
type GitTree <: GitObject
Base.isempty(obj::AbstractGitObject) = (obj.ptr == C_NULL)
```
So, there would be only one `isempty` method.